client2/arq.go

```diff
@@ -0,0 +1,241 @@
+	RecipientQueueID []byte
+
+	// Payload is the message payload
+	Payload []byte
+
+	// SURBID is the SURB identifier.
+	SURBID *[sConstants.SURBIDLength]byte
+
+	// Key is the SURB decryption keys
+	Key []byte
+
+	// Retransmissions counts the number of times the message has been retransmitted.
+	Retransmissions uint32
+
+	// SentAt contains the time the message was sent.
+	SentAt time.Time
+
+	// ReplyETA is the expected round trip time to receive a response.
+	ReplyETA time.Duration
+}
+
+// ARQ is a very simple Automatic Repeat reQuest error correction stategy.
+// Lost packets will be retransmitted. Not an optimized design.
+type ARQ struct {
+	log *log.Logger
+
+	timerQueue *TimerQueue
+	lock       sync.RWMutex
+	gcSurbIDCh chan *[sConstants.SURBIDLength]byte
+	surbIDMap  map[[sConstants.SURBIDLength]byte]*ARQMessage
+
+	sphinxComposerSender SphinxComposerSender
+}
+
+// NewARQ creates a new ARQ.
+func NewARQ(sphinxComposerSender SphinxComposerSender, logbackend io.Writer) *ARQ {
+	log := log.NewWithOptions(logbackend, log.Options{
+		ReportTimestamp: true,
+		Level:           log.DebugLevel,
+		Prefix:          "_ARQ_",
+	})
+	log.Info("NewARQ")
+	return &ARQ{
+		log:                  log,
+		gcSurbIDCh:           make(chan *[sConstants.SURBIDLength]byte),
+		surbIDMap:            make(map[[sConstants.SURBIDLength]byte]*ARQMessage),
+		sphinxComposerSender: sphinxComposerSender,
+	}
+}
+
+// Start starts the ARQ worker thread. You MUST start before using.
+func (a *ARQ) Start() {
+	a.log.Info("Start")
+	a.timerQueue = NewTimerQueue(func(rawSurbID interface{}) {
+		a.log.Info("TimerQueue callback!")
+		surbID, ok := rawSurbID.(*[sConstants.SURBIDLength]byte)
+		if !ok {
+			panic("wtf, failed type assertion!")
+		}
+		a.resend(surbID)
+	})
+	a.timerQueue.Start()
+}
+
+// Stop stops the ARQ's timer queue worker thread.
+func (a *ARQ) Stop() {
+	a.log.Info("Stop")
+	a.timerQueue.Halt()
+	a.timerQueue.Wait()
+}
+
+func (a *ARQ) resend(surbID *[sConstants.SURBIDLength]byte) {
+	a.log.Info("resend")
+
+	a.lock.Lock()
+	message, ok := a.surbIDMap[*surbID]
+	if ok {
+		delete(a.surbIDMap, *surbID)
+		newsurbID := &[sConstants.SURBIDLength]byte{}
+		_, err := rand.Reader.Read(newsurbID[:])
+		if err != nil {
+			panic(err)
+		}
+
+		pkt, k, rtt, err := a.sphinxComposerSender.ComposeSphinxPacket(&Request{
+			WithSURB:          true,
+			SURBID:            newsurbID,
+			DestinationIdHash: message.DestinationIdHash,
+			RecipientQueueID:  message.RecipientQueueID,
+			Payload:           message.Payload,
+			IsSendOp:          true,
+		})
+		if err != nil {
+			panic(err)
+		}
+
+		message.SURBID = newsurbID
+		message.Key = k
+		message.ReplyETA = rtt
+		message.SentAt = time.Now()
+		message.Retransmissions += 1
+
+		a.surbIDMap[*newsurbID] = message
+		priority := uint64(message.SentAt.Add(message.ReplyETA).Add(RoundTripTimeSlop).UnixNano())
+		a.timerQueue.Push(priority, surbID)
+
+		err = a.sphinxComposerSender.SendSphinxPacket(pkt)
```

Comment:
does the lock need to be help for this to work?

Reply:
yes